### PR TITLE
[CF-198] Fix importing of actions

### PR DESCRIPTION
**Pull request by SVilgelm**
Implement module to dynamically load the extensions.
The current implementation is importing modules as the root modules,
does not keep the paths of the modules.

### CODE REVIEW

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.utils.extensions
cloudferrylib/utils/extensions.py:41: [W0602(global-variable-not-assigned), available_extensions] Using global for '_extensions' but no assignment is done
cloudferrylib/utils/extensions.py:42: [W0602(global-variable-not-assigned), available_extensions] Using global for '_extensions_lock' but no assignment is done
************* Module cloudferrylib.scheduler.scenario
cloudferrylib/scheduler/scenario.py:30: [W0102(dangerous-default-value), Scenario.init_tasks] Dangerous default value {} as argument
cloudferrylib/scheduler/scenario.py:51: [W0201(attribute-defined-outside-init), Scenario.init_tasks] Attribute 'tasks' defined outside __init__
cloudferrylib/scheduler/scenario.py:58: [W0201(attribute-defined-outside-init), Scenario.load_scenario] Attribute 'namespace' defined outside __init__
cloudferrylib/scheduler/scenario.py:63: [W0201(attribute-defined-outside-init), Scenario.load_scenario] Attribute 'preparation' defined outside __init__
cloudferrylib/scheduler/scenario.py:60: [W0201(attribute-defined-outside-init), Scenario.load_scenario] Attribute 'migration' defined outside __init__
cloudferrylib/scheduler/scenario.py:66: [W0201(attribute-defined-outside-init), Scenario.load_scenario] Attribute 'rollback' defined outside __init__

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/2501/

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferrylib/utils/extensions.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        def print_log(m):
+            LOG.warning("Cannot import module '%s'", m)
+
+        extensions = []
+        LOG.debug("Scanning available '%s' extensions for %s",
+                  base_class, path or 'all folders')
+        if isinstance(path, basestring):
+            path = [path]
+        else:
+            path = list(path)
```
Comment (roman-verchikov) [inline cloudferrylib/utils/extensions.py]: Looks like even if `path` is already list, you'd still make it list again, like `[[path]]`

File: cloudferrylib/scheduler/scenario.py [hunk 2]
```diff
@@ -32,13 +26,22 @@ class Scenario(object):
...
-    def init_tasks(self, init={}):
+    def init_tasks(self, init):
         with open(self.path_tasks) as tasks_file:
             tasks_file = yaml.load(tasks_file)
-            actions = {}
-            for mod in tasks_file['paths']:
-                actions.update(self.get_actions(mod))
+            # actions = {}
+            # for mod in tasks_file['paths']:
+            #     actions.update(self.get_actions(mod))
```
Comment (roman-verchikov) [inline cloudferrylib/scheduler/scenario.py]: please remove commented out code

File: cloudferrylib/utils/extensions.py (hunk 1 above)
Comment (SVilgelm) [inline cloudferrylib/utils/extensions.py]:
```
>>> list([1,2,3])
[1, 2, 3]
```

File: cloudferrylib/utils/extensions.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                        continue
+                    members = inspect.getmembers(module, inspect.isclass)
+                    for item, klass in members:
+                        if (klass != base_class and
+                                issubclass(klass, base_class)):
+                            LOG.debug("Extension '%s' found!", item)
+                            extensions.append((item, klass))
+
+        if extensions:
+            _extensions[base_class] = extensions
```
Comment (roman-verchikov) [inline cloudferrylib/utils/extensions.py]: If classname matches in different modules, we'll have problems

File: cloudferrylib/scheduler/scenario.py [hunk 4]
```diff
@@ -32,13 +26,22 @@ class Scenario(object):
...
+
+            actions = dict(extensions.available_extensions(
```
Comment (roman-verchikov) [inline cloudferrylib/scheduler/scenario.py]: `extensions.available_extensions` already returns dict, no need to convert it one more time

File: cloudferrylib/scheduler/scenario.py (hunk 2 above)
Comment (SVilgelm) [inline cloudferrylib/scheduler/scenario.py]: Done

File: cloudferrylib/scheduler/scenario.py (hunk 4 above)
Comment (SVilgelm) [inline cloudferrylib/scheduler/scenario.py]: No, It returns a list from tuples: [(name, class), (name, class)]

Comment (roman-verchikov) [inline cloudferrylib/scheduler/scenario.py]: Ah, I see, I'm not right. Maybe return dict there?.. Type conversion to dict looks odd.

Comment (SVilgelm) [inline cloudferrylib/scheduler/scenario.py]: Yep, i agree

File: cloudferrylib/utils/extensions.py (hunk 3 above)
Comment (SVilgelm) [inline cloudferrylib/utils/extensions.py]: What do you mean? Give me an example.

Comment (roman-verchikov) [inline cloudferrylib/utils/extensions.py]: I'm wrong again, thought `base_class` is a string with class name.

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by roman-verchikov** [comment]
+1

**Comment by roman-verchikov** [comment]
can be merged

**Pull request merged**